storage/src/test/java/tech/pegasys/teku/storage/server/kvstore/DatabaseTest.java

```diff
@@ -1666,6 +1666,40 @@ public void shouldStoreFinalizedState(final DatabaseContext context) throws IOEx
     assertThat(database.getLatestAvailableFinalizedState(ONE)).contains(beaconState);
   }
 
+  @TestTemplate
+  public void shouldStoreFinalizedStateAndRoots(final DatabaseContext context) throws IOException {
+    createStorageSystem(context, StateStorageMode.ARCHIVE, StoreConfig.createDefault(), false);
+
+    final BeaconState state0 = dataStructureUtil.randomBeaconState(ZERO);
+    final BeaconState state1 = dataStructureUtil.randomBeaconState(ONE);
+
+    final Bytes32 blockRoot0 = dataStructureUtil.randomBytes32();
+    final Bytes32 blockRoot1 = dataStructureUtil.randomBytes32();
+
+    assertThat(database.getLatestAvailableFinalizedState(ONE)).isEmpty();
+    database.storeFinalizedState(state0, blockRoot0);
+    database.storeFinalizedState(state1, blockRoot1);
+
+    assertThat(database.getLatestAvailableFinalizedState(ZERO)).contains(state0);
+    assertThat(database.getLatestAvailableFinalizedState(ONE)).contains(state1);
+
+    try (final Stream<Map.Entry<Bytes32, UInt64>> roots = database.getFinalizedStateRoots()) {
+      assertThat(getFinalizedStateRootsAtSlot(roots, ZERO)).contains(state0.hashTreeRoot());
+    }
+
+    try (final Stream<Map.Entry<Bytes32, UInt64>> roots = database.getFinalizedStateRoots()) {
+      assertThat(getFinalizedStateRootsAtSlot(roots, ONE)).contains(state1.hashTreeRoot());
+    }
+  }
+
+  private Set<Bytes32> getFinalizedStateRootsAtSlot(
+      final Stream<Map.Entry<Bytes32, UInt64>> roots, final UInt64 slot) {
+    return roots
+        .filter(entry -> slot.equals(entry.getValue()))
+        .map(entry -> entry.getKey())
+        .collect(Collectors.toSet());
+  }
```

Comment:
Probably not super important but given the column is indexed on bytes32, i'd probably change the return to be list, and not bother checking uniqueness, we know its unique already...